src/ui/ui_manager.py

```diff
@@ -1,761 +1,150 @@
-        if h_pot.valid:
-            mouse.move(*h_pot.center, randomize=8, delay_factor=[1.0, 1.5])
-            for _ in range(healing_pots):
-                mouse.click(button="right")
-                wait(0.9, 1.1)
-
-        m_pot = self._template_finder.search_and_wait("SUPER_MANA_POTION", roi=self._config.ui_roi["left_inventory"], time_out=3,normalize_monitor=True)
-        if m_pot.valid is False:  # If not available in shop, try to shop next best potion.
-            m_pot = self._template_finder.search_and_wait("GREATER_MANA_POTION", roi=self._config.ui_roi["left_inventory"], time_out=3,normalize_monitor=True)
-        if m_pot.valid:
-            mouse.move(*m_pot.center, randomize=8, delay_factor=[1.0, 1.5])
-            for _ in range(mana_pots):
-                mouse.click(button="right")
-                wait(0.9, 1.1)
-
+    return False
+
+def detect_screen_object(screen_object, img: np.ndarray = None) -> TemplateMatch:
+    roi = Config().ui_roi[screen_object['roi']] if 'roi' in screen_object else None
+    img = grab() if img is None else img
+    threshold = screen_object['threshold'] if 'threshold' in screen_object else 0.68
+    best_match = screen_object['bestMatch'] if 'bestMatch' in screen_object else False
+    use_grayscale = screen_object['use_grayscale'] if 'use_grayscale' in screen_object else False
+    normalize_monitor = screen_object['normalize_monitor'] if 'normalize_monitor' in screen_object else False
+    match = TemplateFinder().search(
+        ref = screen_object['ref'],
+        inp_img = img,
+        threshold = threshold,
+        roi = roi,
+        best_match = best_match,
+        use_grayscale = use_grayscale,
+        normalize_monitor = normalize_monitor)
+    if match.valid:
+        return match
+    return match
+
+def select_screen_object_match(match: TemplateMatch) -> None:
+    mouse.move(*convert_screen_to_monitor(match.center))
+    wait(0.05, 0.09)
+    mouse.click("left")
+    wait(0.05, 0.09)
+
+def wait_for_screen_object(screen_object, time_out: int = None) -> TemplateMatch:
+    roi = Config().ui_roi[screen_object['roi']] if 'roi' in screen_object else None
+    threshold = screen_object['threshold'] if 'threshold' in screen_object else 0.68
+    best_match = screen_object['bestMatch'] if 'bestMatch' in screen_object else False
+    use_grayscale = screen_object['use_grayscale'] if 'use_grayscale' in screen_object else False
+    normalize_monitor = screen_object['normalize_monitor'] if 'normalize_monitor' in screen_object else False
+    time_out = time_out if time_out else 30
+    match = TemplateFinder().search_and_wait(
+        ref = screen_object['ref'],
+        time_out = time_out,
+        threshold = threshold,
+        roi = roi,
+        best_match = best_match,
+        use_grayscale = use_grayscale,
+        normalize_monitor = normalize_monitor)
+    if match.valid:
+        return match
+    return match
```

Comment:
Already mentioned to Ezro, but lets cleanup the redundant code here before merging to master. Whether that means making a @dataclass for the default values, getting rid of template_finder.search_and_wait and just making that a part of ui_manager.py, or whatever it is.

Reply:
This should be resolved with the introduction of ScreenObject / ScreenObjects.